pkg/reconcile/common.go

```diff
@@ -0,0 +1,159 @@
+package reconcileCommon
+
+import (
+	"context"
+	"reflect"
+
+	"github.com/openshift/route-monitor-operator/pkg/util/finalizer"
+	"github.com/openshift/route-monitor-operator/pkg/util/reconcile"
+
+	configv1 "github.com/openshift/api/config/v1"
+	"github.com/openshift/route-monitor-operator/api/v1alpha1"
+	customerrors "github.com/openshift/route-monitor-operator/pkg/util/errors"
+	monitoringv1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+//go:generate mockgen -source $GOFILE -destination ../util/test/generated/mocks/reconcile/common.go -package $GOPACKAGE
+
+type ResourceComparerInterface interface {
+	DeepEqual(x, y interface{}) bool
+}
+
+type ResourceComparer struct{}
+
+func (_ *ResourceComparer) DeepEqual(x, y interface{}) bool {
+	return reflect.DeepEqual(x, y)
+}
+
+type MonitorReconcileCommon struct {
+	Client   client.Client
+	Ctx      context.Context
+	Comparer ResourceComparerInterface
+}
+
+func NewMonitorReconcileCommon(ctx context.Context, c client.Client) *MonitorReconcileCommon {
+	return &MonitorReconcileCommon{
+		Client:   c,
+		Ctx:      ctx,
+		Comparer: &ResourceComparer{},
+	}
+}
+
+// returns whether the errorStatus has changed
+func (u *MonitorReconcileCommon) SetErrorStatus(errorStatus *string, err error) bool {
+	switch {
+	case u.areErrorAndErrorStatusFull(errorStatus, err):
+		return false
+	case u.needsErrorStatusToBeFlushed(errorStatus, err):
+		*errorStatus = ""
+		return true
+	case u.needsErrorStatusToBeSet(errorStatus, err):
+		*errorStatus = err.Error()
+		return true
+	}
+	return false
+}
+
+// If an error has already been flagged and still occurs
+func (u *MonitorReconcileCommon) areErrorAndErrorStatusFull(errorStatus *string, err error) bool {
+	return *errorStatus != "" && err != nil
+}
+
+// If the error was flagged but stopped firing
+func (u *MonitorReconcileCommon) needsErrorStatusToBeFlushed(errorStatus *string, err error) bool {
+	return *errorStatus != "" && err == nil
+}
+
+// If the error was not flagged but has started firing
+func (u *MonitorReconcileCommon) needsErrorStatusToBeSet(errorStatus *string, err error) bool {
+	return *errorStatus == "" && err != nil
+}
+
+func (u *MonitorReconcileCommon) SetResourceReference(reference *v1alpha1.NamespacedName, targetNamespace types.NamespacedName) (bool, error) {
+	desiredRef := v1alpha1.NamespacedName{Name: targetNamespace.Name, Namespace: targetNamespace.Namespace}
+	if *reference == (v1alpha1.NamespacedName{}) {
+		*reference = desiredRef
+		return true, nil
+	} else if desiredRef == (v1alpha1.NamespacedName{}) {
+		*reference = desiredRef
+		return true, nil
+	} else if *reference != desiredRef {
+		// TODO Check when this is really required
+		return false, customerrors.InvalidReferenceUpdate
```

Comment:
This case happens when a URLMonitor/RouteMonitor with existing ServiceMinorRef is moved from namespace to namespace. In this case, the reconciler tries to update the existing reference with the ServiceMonitor created in the namespace, but we don't yet support the movement or alterings of already existing sub-resources. 
It's also not really required I would say. In my opinion, we could just create another ServiceMonitor and replace the existing reference. At the moment we are just constantly re-queueing with an error. I am not sure if this is the best way to go.
cc @sulmer-redhat . I think I extracted that part from your code in [here](https://github.com/openshift/route-monitor-operator/blob/e2c7b17d283b95e007f68538bd6b5d7f37e5eec5/controllers/routemonitor/adder/adder.go#L83) Am I wrong about it?